galley/pkg/config/analysis/analyzers/virtualservice/destinationhosts.go

```diff
@@ -0,0 +1,102 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package virtualservice
+
+import (
+	"istio.io/api/networking/v1alpha3"
+	"istio.io/istio/galley/pkg/config/analysis"
+	"istio.io/istio/galley/pkg/config/analysis/analyzers/util"
+	"istio.io/istio/galley/pkg/config/analysis/msg"
+	"istio.io/istio/galley/pkg/config/collection"
+	"istio.io/istio/galley/pkg/config/processor/metadata"
+	"istio.io/istio/galley/pkg/config/resource"
+)
+
+// DestinationHostAnalyzer checks the destination hosts associated with each virtual service
+type DestinationHostAnalyzer struct{}
+
+var _ analysis.Analyzer = &DestinationHostAnalyzer{}
+
+type hostAndSubset struct {
+	host   resource.Name
+	subset string
+}
+
+// Metadata implements Analyzer
+func (d *DestinationHostAnalyzer) Metadata() analysis.Metadata {
+	return analysis.Metadata{
+		Name: "virtualservice.DestinationHostAnalyzer",
+		Inputs: collection.Names{
+			metadata.IstioNetworkingV1Alpha3SyntheticServiceentries,
+			metadata.IstioNetworkingV1Alpha3Serviceentries,
+			metadata.IstioNetworkingV1Alpha3Virtualservices,
+		},
+	}
+}
+
+// Analyze implements Analyzer
+func (d *DestinationHostAnalyzer) Analyze(ctx analysis.Context) {
+	// Precompute the set of service entry hosts that exist (there can be more than one defined per ServiceEntry CRD)
+	serviceEntryHosts := initServiceEntryHostNames(ctx)
+
+	ctx.ForEach(metadata.IstioNetworkingV1Alpha3Virtualservices, func(r *resource.Entry) bool {
+		d.analyzeVirtualService(r, ctx, serviceEntryHosts)
+		return true
+	})
+}
+
+func (d *DestinationHostAnalyzer) analyzeVirtualService(r *resource.Entry, ctx analysis.Context,
+	serviceEntryHosts map[resource.Name]bool) {
+
+	vs := r.Item.(*v1alpha3.VirtualService)
+	ns, _ := r.Metadata.Name.InterpretAsNamespaceAndName()
+
+	destinations := getRouteDestinations(vs)
+
+	for _, destination := range destinations {
+		if !d.checkDestinationHost(ns, destination, ctx, serviceEntryHosts) {
+			ctx.Report(metadata.IstioNetworkingV1Alpha3Virtualservices,
+				msg.NewReferencedResourceNotFound(r, "host", destination.GetHost()))
+		}
+	}
+}
+
+func (d *DestinationHostAnalyzer) checkDestinationHost(vsNamespace string, destination *v1alpha3.Destination,
+	ctx analysis.Context, serviceEntryHosts map[resource.Name]bool) bool {
+
+	name := util.GetResourceNameFromHost(vsNamespace, destination.GetHost())
+
+	// Check explicitly defined ServiceEntries as well as services discovered from the platform
+	if _, ok := serviceEntryHosts[name]; ok {
+		return true
+	}
+	if ctx.Exists(metadata.IstioNetworkingV1Alpha3SyntheticServiceentries, name) {
+		return true
+	}
+
+	return false
+}
+
+func initServiceEntryHostNames(ctx analysis.Context) map[resource.Name]bool {
```

Comment:
Would it make sense to actively put such methods under a common folder? If nothing else, it will help with single-instancing the code and logic for various analyzers that might need something similar. Then, introducing re-use of cached/index data would be simpler.

Reply:
I like the idea of gathering "pre-compute" methods in one place.